public/app/plugins/datasource/grafana-pyroscope-datasource/QueryEditor/QueryEditor.tsx

```diff
@@ -142,28 +148,28 @@ function useLabels(
   useEffect(() => {
     const fetchData = async () => {
       const labels = await datasource.getLabelNames(
-        createSelector(query.labelSelector, query.profileTypeId, ''),
+        createSelector(queryRef.current.labelSelector, queryRef.current.profileTypeId, ''),
         unpreciseRange.from,
         unpreciseRange.to
       );
 
       setLabels(labels);
     };
     fetchData();
-  }, [query, unpreciseRange.from, unpreciseRange.to, datasource, setLabels]);
+  }, [queryRef, unpreciseRange.from, unpreciseRange.to, datasource, setLabels]);
 
   // Create a function with range and query already baked in, so we don't have to send those everywhere
   const getLabelValues = useCallback(
     (label: string) => {
-      let labelSelector = createSelector(query.labelSelector, query.profileTypeId, label);
+      const labelSelector = createSelector(queryRef.current.labelSelector, queryRef.current.profileTypeId, label);
       return datasource.getLabelValues(labelSelector, label, unpreciseRange.from, unpreciseRange.to);
     },
-    [datasource, query.labelSelector, query.profileTypeId, unpreciseRange.to, unpreciseRange.from]
+    [queryRef, datasource, unpreciseRange.from, unpreciseRange.to]
```

Comment:
I don't think we need to change these callbacks. When the query changes the callback should be recreated with the newest query and there should not be an issue here.